src/Eloquent/Builder.php

```diff
@@ -183,6 +186,41 @@ public function raw($value = null)
         return $results;
     }
 
+    /**
+     * Attempt to create the record if it does not exist with the matching attributes.
+     * If the record exists, it will be returned.
+     *
+     * @param  array $attributes The attributes to check for duplicate records
+     * @param  array $values     The attributes to insert if no matching record is found
+     */
+    public function createOrFirst(array $attributes = [], array $values = []): Model
+    {
+        // Apply casting and default values to the attributes
+        $instance = $this->newModelInstance($values + $attributes);
```

Comment:
When doing `$values + $attributes`, preference is given to an existing key in `$values`. In that case, it seems possible that actual criteria in `$attributes` might be lost.

Is there a particular reason you didn't do `$attributes + $values`?

Reply:
This sounds like a possible Laravel bug if you want to pursue that.